napari/utils/misc.py

```diff
@@ -267,6 +267,16 @@ def __str__(self):
         """
         return self.value
 
+    def __eq__(self, other):
+        if type(self) is type(other):
+            return self is other
+        elif isinstance(other, str):
+            return str(self) == other
+        return NotImplemented
+
+    def __hash__(self):
```

Comment:
perhaps use `id(self)` instead?  for the same reason you mentioned in your post, do we want `SomeEnum.SOMEVALUE` and `AnotherEnum.SOMEVALUE` to overwrite each other when used as keys in a dict, or items in a set?

Reply:
The way I think about `__hash__` is that it's a shortcut for checking inequality. I.e. if two objects have different hash values, then they are not equal and we don't actually need to call `__eq__` to check that. However, if they have equal hashes, they are not necessarily equal. This is why when you override `__eq__` (i.e. you change the definition of equality), you must also override `__hash__`.

For `set`, the equal hashes of `SomeEnum.SOMEVALUE` and `AnotherEnum.SOMEVALUE`, mean that there will be hash collision (i.e. the hash values are equal), but this only leads to a true equality check (i.e. a call to `__eq__`), in which case the values of `SomeEnum.SOMEVALUE` and `AnotherEnum.SOMEVALUE` are not equal and will not overwrite each other.

I added some tests to cover this logic:

```python
    assert TestEnum.THING in {TestEnum.THING, TestEnum.OTHERTHING}
    assert TestEnum.THING not in {TestEnum.OTHERTHING}
    assert TestEnum.THING in {'thing', TestEnum.OTHERTHING}
    assert TestEnum.THING not in {AnotherTestEnum.THING, AnotherTestEnum.ANOTHERTHING}
```

Note that `TestEnum.THING in {'thing', TestEnum.OTHERTHING}` is `True` because that's how we defined equality.

If we used `id(self)` instead of `hash(str(self))`, then the definitions of `__eq__` and `__hash__` would be contradictory because `hash(TestEnum.THING) != hash('thing')` but `TestEnum.THING == 'thing')`.

Reply:
thanks again for the clear explanation!  what I was more worried about was `{TestEnum.THING, AnotherTestEnum.THING}` returning a set with a single item, but I hadn't appreciated the fallback to `__eq__` here.